Lib/asyncio/staggered.py

```diff
@@ -61,78 +61,67 @@ async def staggered_race(coro_fns, delay, *, loop=None):
-            # task as well. If we do so, then since there is no `await` after
-            # here and CancelledError are usually thrown at one, we will
-            # encounter a curious corner case where the current task will end
-            # up as done() == True, cancelled() == False, exception() ==
-            # asyncio.CancelledError. This behavior is specified in
-            # https://bugs.python.org/issue30048
-            for i, t in enumerate(running_tasks):
-                if i != this_index:
-                    t.cancel()
-
-    first_task = loop.create_task(run_one_coro(None))
-    running_tasks.append(first_task)
-    try:
-        # Wait for a growing list of tasks to all finish: poor man's version of
-        # curio's TaskGroup or trio's nursery
-        done_count = 0
-        while done_count != len(running_tasks):
-            done, _ = await tasks.wait(running_tasks)
-            done_count = len(done)
-            # If run_one_coro raises an unhandled exception, it's probably a
-            # programming error, and I want to see it.
-            if __debug__:
-                for d in done:
-                    if d.done() and not d.cancelled() and d.exception():
-                        raise d.exception()
-        return winner_result, winner_index, exceptions
-    finally:
-        # Make sure no tasks are left running if we leave this function
-        for t in running_tasks:
-            t.cancel()
+            if error is not None:
+                return
+
+        nonlocal winner_result, winner_index
+        # If this is in an eager task factory, it's possible
+        # for multiple tasks to get here. In that case, we want
+        # only the first one to win and the rest to no-op before
+        # cancellation.
+        if winner_result is None and not task_group._aborting:
+            winner_result = future.result()
+            winner_index = index
+
+            # Cancel all other tasks, we win!
+            task_group._abort()
+
+    async with taskgroups.TaskGroup() as task_group:
+        for index, coro in enumerate(coro_fns):
+            if task_group._aborting:
+                break
+
+            exceptions.append(None)
+            task = loop.create_task(coro())
+
+            # We don't want the task group to propagate the error. Instead,
+            # we want to put it in our special exceptions list, so we manually
+            # create the task.
+            task.add_done_callback(task_group._on_task_done_without_propagation)
+            task_group._tasks.add(task)
+
+            # We need this extra wrapper here to stop the closure from having
+            # an incorrect index.
+            def wrapper(idx):
+                return lambda future: future_callback(idx, future, task_group)
+
+            task.add_done_callback(wrapper(index))
+
+            if delay is not None:
+                await tasks.sleep(delay or 0)
```

Comment:
You need to wait for the previous task to fail or the timeout to expire
```python
with contextlib.suppress(TimeoutError):
    await asyncio.wait_for(previous_task.wait(), delay)
```

Reply:
There probably needs to be a test for this feature

Reply:
There is, see `Lib/test/test_asyncio/test_staggered.py`.

Reply:
> You need to wait for the previous task to fail or the timeout to expire
> 
> 

It should be doing this right now, but maybe I implemented the timeouts incorrectly? This should only create the task after the timeout expires. (Though, the staggering tests should have failed, I'll double check this.)

Reply:
Doesn't look like it's tested: 

```python
    async def test_first_error_cancels_delay_second_successful(self):
        async def coro(index):
            if index == 0:
                raise ValueError
            return f'Res: {index}'
        loop = asyncio.get_running_loop()
        start = loop.time()
        winner, index, excs = await staggered_race(
            [
                lambda: coro(0),
                lambda: coro(1),
            ],
            delay=10,
        )
        stop = loop.time()
        self.assertEqual(winner, 'Res: 2')
        self.assertLess(stop - start, 1)
```

Reply:
We have one test with a delay and `sleep`:

```py
    async def test_first_timeout_second_successful(self):
        async def coro(index):
            if index == 0:
                await asyncio.sleep(10)  # much bigger than delay
            return f'Res: {index}'

        winner, index, excs = await staggered_race(
            [
                lambda: coro(0),
                lambda: coro(1),
            ],
            delay=0.1,
        )

        self.assertEqual(winner, 'Res: 1')
        self.assertEqual(index, 1)
        self.assertEqual(len(excs), 2)
        self.assertIsInstance(excs[0], asyncio.CancelledError)
        self.assertIs(excs[1], None)
```

But I agree, that isn't nearly enough. I'll add another.

Reply:
This is the short delay with long timeout, I'm interested in the long delay and early failure